# Unix - Conda - Pip

We run commands in CLI, e.g., Terminal, but here we are using Jupyter Notebook to demonstrate:

1. Output from commands
2. That when experiementing or developing programs in the notebook, you can avoid switching tools.
3. That you can install pacakges in the beginning of the notebook (e.g., Sagemaker) so it's clear what packages are used.

## Unix commands

### Basics

For some of the basic commands, you can run use the cell to run it directly:

In [1]:
pwd # print working directory

'/mnt/f/4thbr/code/4bmle9/software-dev-for-ml-101/notebooks'

You always want to know "where you are" and how to look for help. Use `man` to display the user manual of any command that we can run on the terminal. 

In [2]:
man pwd # press ESC to exit

Further, cell magic `%%bash` turns a cell into a bash script where you can run multiple lines (See more cell/line magics [here](https://ipython.readthedocs.io/en/stable/interactive/magics.html)). 

In [3]:
%%bash
pwd
ls -lah # WE DO: What are those flags for?

# fsl
# ls flags:
# -l long format including permissions
# -h (with -l flag) human readable file sizes
# -a show hidden files including those beginning with '.'

/mnt/f/4thbr/code/4bmle9/software-dev-for-ml-101/notebooks
total 40K
drwxr-xr-x 1 fsl fsl 512 Aug 10 13:40 .
drwxr-xr-x 1 fsl fsl 512 Aug 10 12:51 ..
drwxr-xr-x 1 fsl fsl 512 Aug 10 13:38 .ipynb_checkpoints
-rw-r--r-- 1 fsl fsl 37K Aug 10 13:40 unix-conda-pip.ipynb


### Download a data file

Let's download some data from [Census Income Data Set](https://archive.ics.uci.edu/ml/datasets/Census+Income); but first we would like to make a new directory to store the data at the same level as `nb` directory.

In [5]:
%%bash
cd ..
# WE DO: make a new diretory called `dat` under the project root directory
pwd
ls -lah
mkdir dat


/mnt/f/4thbr/code/4bmle9/software-dev-for-ml-101
total 20K
drwxr-xr-x 1 fsl fsl 512 Aug 10 12:51 .
drwxr-xr-x 1 fsl fsl 512 Aug 10 13:06 ..
drwxr-xr-x 1 fsl fsl 512 Aug 10 13:48 notebooks
-rw-r--r-- 1 fsl fsl 17K Aug 10 12:51 software-dev-for-ml-101.md


Let's confirm that we just created an empty directory `dat` using command `tree`. Not all command can be run in the cell directly, adding ! in front of command can do the trick:

In [8]:
%%bash
cd ..
pwd
ls -lah

/mnt/f/4thbr/code/4bmle9/software-dev-for-ml-101
total 20K
drwxr-xr-x 1 fsl fsl 512 Aug 10 13:48 .
drwxr-xr-x 1 fsl fsl 512 Aug 10 13:06 ..
drwxr-xr-x 1 fsl fsl 512 Aug 10 13:48 dat
drwxr-xr-x 1 fsl fsl 512 Aug 10 13:58 notebooks
-rw-r--r-- 1 fsl fsl 17K Aug 10 12:51 software-dev-for-ml-101.md


In [9]:
!tree .. -L 2

/bin/bash: tree: command not found


In [10]:
# download a csv file
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data -O ../dat/adult.csv

--2022-08-10 13:59:56--  https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 3974305 (3.8M) [application/x-httpd-php]
Saving to: ‘../dat/adult.csv’

../dat/adult.csv    100%[===================>]   3.79M  6.48MB/s    in 0.6s    

2022-08-10 13:59:57 (6.48 MB/s) - ‘../dat/adult.csv’ saved [3974305/3974305]



In [7]:
# (fsl commented out this code for output comparison)
# !tree .. -L 2 # check again

..
├── LICENSE
├── README.md
├── dat
│   └── adult.csv
├── md
│   └── git-more.md
└── nb
    └── unix-conda-pip.ipynb

3 directories, 5 files


In [7]:
!tree .. -L 2 # check again

..
├── LICENSE
├── README.md
├── dat
│   └── adult.csv
├── md
│   └── git-more.md
└── nb
    └── unix-conda-pip.ipynb

3 directories, 5 files


### Inspect data

Let's inspect first 10 lines of data:

In [11]:
!head ../dat/adult.csv

39, State-gov, 77516, Bachelors, 13, Never-married, Adm-clerical, Not-in-family, White, Male, 2174, 0, 40, United-States, <=50K
50, Self-emp-not-inc, 83311, Bachelors, 13, Married-civ-spouse, Exec-managerial, Husband, White, Male, 0, 0, 13, United-States, <=50K
38, Private, 215646, HS-grad, 9, Divorced, Handlers-cleaners, Not-in-family, White, Male, 0, 0, 40, United-States, <=50K
53, Private, 234721, 11th, 7, Married-civ-spouse, Handlers-cleaners, Husband, Black, Male, 0, 0, 40, United-States, <=50K
28, Private, 338409, Bachelors, 13, Married-civ-spouse, Prof-specialty, Wife, Black, Female, 0, 0, 40, Cuba, <=50K
37, Private, 284582, Masters, 14, Married-civ-spouse, Exec-managerial, Wife, White, Female, 0, 0, 40, United-States, <=50K
49, Private, 160187, 9th, 5, Married-spouse-absent, Other-service, Not-in-family, Black, Female, 0, 0, 16, Jamaica, <=50K
52, Self-emp-not-inc, 209642, HS-grad, 9, Married-civ-spouse, Exec-managerial, Husband, White, Male, 0, 0, 45, United-States, >50K
31, 

How about the last 3 lines? ( wait, I only see 2 lines. what's wrong? ) 

In [12]:
!tail -n 3 ../dat/adult.csv

22, Private, 201490, HS-grad, 9, Never-married, Adm-clerical, Own-child, White, Male, 0, 0, 20, United-States, <=50K
52, Self-emp-inc, 287927, HS-grad, 9, Married-civ-spouse, Exec-managerial, Wife, White, Female, 15024, 0, 40, United-States, >50K



How many records are there?

In [13]:
!wc -l ../dat/adult.csv

32562 ../dat/adult.csv


Challenge: how many columns are there?

In [14]:
!head -1 ../dat/adult.csv | sed 's/[^,]//g' | wc -c

15


## Conda: environment manager + package manager

You might have heard about Anaconda / Miniconda / Miniforge:

- What is the difference between `conda` and `miniconda`? [An answer](https://stackoverflow.com/questions/45421163/anaconda-vs-miniconda).
- What is the difference between `miniconda` and `miniforge`? [An answer](https://stackoverflow.com/questions/60532678/what-is-the-difference-between-miniconda-and-miniforge).

Though Conda is considered both a package manager and an environment manager, we  focus on using it as the environment manager, and use `pip` as the package manager.

### Installation

First, install `miniconda` (Feel free to skip if you have already installed it, or any flavor of conda).

Run the following commands in terminal to download the lastest Miniconda distribution and install it (Mac Intel Book). 

    ```
    cd ~
    wget https://repo.anaconda.com/miniconda/Miniconda3-latest-MacOSX-x86_64.sh
    bash Miniconda3-latest-MacOSX-x86_64.sh
    ```

For M1 chip, install `miniforge` instead. If you are new to `brew`, following instructions [here](https://brew.sh) to install Homebrew first:

    ```
    brew install miniforge
    ```

Outputs from these installations are rather long, so we ask you to run it in a terminal window instead.

Confirm that the installation is successful:

In [1]:
# (fsl commented out this code for output comparison)
# %%bash
# which conda  # where the executable conda is in my path?
# conda -V     # which version of conda?
# which python # where the executable python is in my path?
# python -V    # what version of python?

/usr/local/Caskroom/miniforge/base/bin/conda
conda 4.12.0
/usr/local/Caskroom/miniforge/base/bin/python
Python 3.9.10


In [15]:
%%bash
which conda  # where the executable conda is in my path?
conda -V     # which version of conda?
which python # where the executable python is in my path?
python -V    # what version of python?

/home/fsl/miniconda3/envs/mle9/bin/conda
conda 4.13.0
/home/fsl/miniconda3/envs/mle9/bin/python
Python 3.9.7


### Create a new env

Create a new conda environment named `py39_12` where you specify a python version to be 3.9.12. Note the last flag `--yes` to skip the confirmation prompt.

In [2]:
# (fsl comment this code out for output comparison)
# !conda create --name py39_12 python=3.9.12 --yes

Solving environment: done

## Package Plan ##

  environment location: /usr/local/Caskroom/miniforge/base/envs/py39_12

  added / updated specs:
    - python=3.9.12


The following NEW packages will be INSTALLED:

  bzip2              conda-forge/osx-64::bzip2-1.0.8-h0d85af4_4
  ca-certificates    conda-forge/osx-64::ca-certificates-2021.10.8-h033912b_0
  libffi             conda-forge/osx-64::libffi-3.4.2-h0d85af4_5
  libzlib            conda-forge/osx-64::libzlib-1.2.11-h6c3fc93_1014
  ncurses            conda-forge/osx-64::ncurses-6.3-h96cf925_1
  openssl            conda-forge/osx-64::openssl-3.0.2-h6c3fc93_1
  pip                conda-forge/noarch::pip-22.0.4-pyhd8ed1ab_0
  python             conda-forge/osx-64::python-3.9.12-h1cc4136_1_cpython
  python_abi         conda-forge/osx-64::python_abi-3.9-2_cp39
  readline           conda-forge/osx-64::readline-8.1-h05e3726_0
  setuptools         conda-forge/osx-64::setuptools-62.1.0-py39h6e9494a_0
  sqlite             conda-forge/osx-6

In [16]:
!conda create --name py39_12 python=3.9.12 --yes

Solving environment: done

## Package Plan ##

  environment location: /home/fsl/miniconda3/envs/mle9/envs/py39_12

  added / updated specs:
    - python=3.9.12


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    _libgcc_mutex-0.1          |             main           3 KB
    _openmp_mutex-5.1          |            1_gnu          21 KB
    ca-certificates-2022.07.19 |       h06a4308_0         124 KB
    certifi-2022.6.15          |   py39h06a4308_0         153 KB
    ld_impl_linux-64-2.38      |       h1181459_1         654 KB
    libffi-3.3                 |       he6710b0_2          50 KB
    libgcc-ng-11.2.0           |       h1234567_1         5.3 MB
    libgomp-11.2.0             |       h1234567_1         474 KB
    libstdcxx-ng-11.2.0        |       h1234567_1         4.7 MB
    ncurses-6.3                |       h5eee18b_3         781 KB
    openssl-1.1.1q             |       h7f8

In [17]:
# YOU CODE HERE 
# list all environments in conda
!conda env list

# conda environments:
#
                         /home/fsl/miniconda3
                         /home/fsl/miniconda3/envs/ibm
                         /home/fsl/miniconda3/envs/ibmex
                         /home/fsl/miniconda3/envs/jpu38
base                  *  /home/fsl/miniconda3/envs/mle9
py39_12                  /home/fsl/miniconda3/envs/mle9/envs/py39_12



You shall see something like this:
```
# conda environments:
#
                         /Users/flora/miniforge3
                         /Users/flora/miniforge3/envs/tf38
base                  *  /usr/local/Caskroom/miniforge/base
py39                     /usr/local/Caskroom/miniforge/base/envs/py39
py39_12                  /usr/local/Caskroom/miniforge/base/envs/py39_12
```

### Activate an env

By default, you are under the `base` environment, to activate the new environement, run `conda activate py39_12` in terminal.

If you are using VS code, click on the top right button to switch python kernel. You will be prompted to install `ipykernel`, follow the instructions to install the package. Or you can run `source activate`, followed by 
`conda activate /usr/local/Caskroom/miniforge/base/envs/py39_12` in a cell. 

If you are using jupyter notebook on a localhost (by defaut 8888), you can restart the jupyter notebook after activating the new environment.

Now verify the python version:

In [1]:
# (fsl comment out code to preserve output for comparison)
# %%bash
# 
# python --version
# which python
# conda env list

Python 3.9.12
/usr/local/Caskroom/miniforge/base/envs/py39_12/bin/python
# conda environments:
#
                         /Users/flora/miniforge3
                         /Users/flora/miniforge3/envs/tf38
base                     /usr/local/Caskroom/miniforge/base
py39                     /usr/local/Caskroom/miniforge/base/envs/py39
py39_12               *  /usr/local/Caskroom/miniforge/base/envs/py39_12



In [18]:
%%bash

python --version
which python
conda env list

Python 3.9.7
/home/fsl/miniconda3/envs/mle9/bin/python
# conda environments:
#
                         /home/fsl/miniconda3
                         /home/fsl/miniconda3/envs/ibm
                         /home/fsl/miniconda3/envs/ibmex
                         /home/fsl/miniconda3/envs/jpu38
base                  *  /home/fsl/miniconda3/envs/mle9
py39_12                  /home/fsl/miniconda3/envs/mle9/envs/py39_12



<details>
<summary>Click to see how to add the Conda env to you jupyter notebook kernels</summary>

```
pip install ipykernel
python -m ipykernel install --user --name=py39_12
```

Next time you launch jupyter notebook, you will see `py39_12` as an option under Kernel/Change kernel.
</details>




### Delete an env

To keep a lean list of environments, we want to prune unused environements from time to time. Simply do the following
```
conda remove --name old_env --all --yes
```

## Pip: python package manager

We recommend to use pip as your python package installer/manager [(fun read on pip vs conda.)](https://stackoverflow.com/questions/20994716/what-is-the-difference-between-pip-and-conda) Note you don't have install pip explictly since it was done during the installation of miniconda/miniforge. 

In [2]:
# (fsl comment out code to preserve output for comparison)
# !which pip

/usr/local/Caskroom/miniforge/base/envs/py39_12/bin/pip


In [20]:
!which pip
!pip --version

/home/fsl/miniconda3/envs/mle9/bin/pip
pip 21.2.4 from /home/fsl/miniconda3/envs/mle9/lib/python3.9/site-packages/pip (python 3.9)


### Install package

We can certainly run `pip` commands the same way as earlier. Just for fun, let's try Jupyter line magic `%pip`. 

First make sure that we would like to have the most recent version of `pip` installed.

In [ ]:
# fsl inserted cell below to verify current active env in which pip will be 
# upgraded and numpy/pandas installed (see cells +2, +3 below)

In [22]:
%%bash
python --version
which python
conda env list

Python 3.9.7
/home/fsl/miniconda3/envs/mle9/bin/python
# conda environments:
#
                         /home/fsl/miniconda3
                         /home/fsl/miniconda3/envs/ibm
                         /home/fsl/miniconda3/envs/ibmex
                         /home/fsl/miniconda3/envs/jpu38
base                  *  /home/fsl/miniconda3/envs/mle9
py39_12                  /home/fsl/miniconda3/envs/mle9/envs/py39_12



In [24]:
%%bash
which pip
pip --version

which ipython
ipython --version

conda list

/home/fsl/miniconda3/envs/mle9/bin/pip
pip 21.2.4 from /home/fsl/miniconda3/envs/mle9/lib/python3.9/site-packages/pip (python 3.9)
/home/fsl/miniconda3/envs/mle9/bin/ipython
7.29.0
# packages in environment at /home/fsl/miniconda3/envs/mle9:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                        main  
_openmp_mutex             4.5                       1_gnu  
alabaster                 0.7.12             pyhd3eb1b0_0  
anaconda                  2021.11                  py39_0  
anaconda-client           1.9.0            py39h06a4308_0  
anaconda-project          0.10.1             pyhd3eb1b0_0  
anyio                     2.2.0            py39h06a4308_1  
appdirs                   1.4.4              pyhd3eb1b0_0  
argh                      0.26.2           py39h06a4308_0  
argon2-cffi               20.1.0           py39h27cfd23_1  
arrow                     0.13.1           py39h06a4308_0  
asn1crypto                1.4

In [34]:
%%bash
which conda
conda --version
which numpy
numpy --version
which pandas
pandas --version
conda list numpy
conda list pandas

/home/fsl/miniconda3/envs/mle9/bin/conda
conda 4.13.0
# packages in environment at /home/fsl/miniconda3/envs/mle9:
#
# Name                    Version                   Build  Channel
numpy                     1.20.3           py39hf144106_0  
numpy-base                1.20.3           py39h74d4b33_0  
numpydoc                  1.1.0              pyhd3eb1b0_1  
# packages in environment at /home/fsl/miniconda3/envs/mle9:
#
# Name                    Version                   Build  Channel
pandas                    1.3.4            py39h8c16a72_0  


bash: line 4: numpy: command not found
bash: line 6: pandas: command not found


### (fsl) skipping pip-upgrading pip below since just created conda mle9 env using miniconda  
(with anaconda = 2021.11)

In [3]:
# %pip install --upgrade pip

Note: you may need to restart the kernel to use updated packages.


### (fsl) also skipping pip-installing numpy, pandas since just created conda mle9 env using miniconda
(with anaconda = 2021.11)

In [4]:
# %pip install numpy pandas

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.7/12.7 MB 6.3 MB/s eta 0:00:0000:0100:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.9/9.9 MB 6.1 MB/s eta 0:00:0000:0100:01
Note: you may need to restart the kernel to use updated packages.


Now you can use these packages:

# Allowing multiple outputs to a single cell (USEFUL!)

In [32]:
# allow multiple outputs in a single cell
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

In [27]:
import numpy as np

# create a numpy array with random numbers
x = np.random.rand(10)
x.shape
x.reshape((5,-1))
print('x = ', x)

x =  [0.17413963 0.01740868 0.01609784 0.67666189 0.01322551 0.46642191
 0.26925339 0.69823748 0.62945049 0.99911869]


In [28]:
%precision 3
x.reshape((5,-1))

array([[0.174, 0.017],
       [0.016, 0.677],
       [0.013, 0.466],
       [0.269, 0.698],
       [0.629, 0.999]])

In [33]:
import pandas as pd
census_df = pd.read_csv(filepath_or_buffer='../dat/adult.csv', header = None)
census_df.head()
census_df.tail()
census_df.shape

0                  1       2           3   4                    5   \
0  39          State-gov   77516   Bachelors  13        Never-married   
1  50   Self-emp-not-inc   83311   Bachelors  13   Married-civ-spouse   
2  38            Private  215646     HS-grad   9             Divorced   
3  53            Private  234721        11th   7   Married-civ-spouse   
4  28            Private  338409   Bachelors  13   Married-civ-spouse   

                   6               7       8        9     10  11  12  \
0        Adm-clerical   Not-in-family   White     Male  2174   0  40   
1     Exec-managerial         Husband   White     Male     0   0  13   
2   Handlers-cleaners   Not-in-family   White     Male     0   0  40   
3   Handlers-cleaners         Husband   Black     Male     0   0  40   
4      Prof-specialty            Wife   Black   Female     0   0  40   

               13      14  
0   United-States   <=50K  
1   United-States   <=50K  
2   United-States   <=50K  
3   United-States   <=50K  
4            Cuba   <=50K

0              1       2            3   4                    5   \
32556  27        Private  257302   Assoc-acdm  12   Married-civ-spouse   
32557  40        Private  154374      HS-grad   9   Married-civ-spouse   
32558  58        Private  151910      HS-grad   9              Widowed   
32559  22        Private  201490      HS-grad   9        Never-married   
32560  52   Self-emp-inc  287927      HS-grad   9   Married-civ-spouse   

                       6           7       8        9      10  11  12  \
32556        Tech-support        Wife   White   Female      0   0  38   
32557   Machine-op-inspct     Husband   White     Male      0   0  40   
32558        Adm-clerical   Unmarried   White   Female      0   0  40   
32559        Adm-clerical   Own-child   White     Male      0   0  20   
32560     Exec-managerial        Wife   White   Female  15024   0  40   

                   13      14  
32556   United-States   <=50K  
32557   United-States    >50K  
32558   United-States   <=50K  
32559   United-States   <=50K  
32560   United-States    >50K

(32561, 15)

### Uninstall packages

In [9]:
# Uninstall a package
# %pip uninstall numpy pandas --yes

Found existing installation: numpy 1.22.3
Uninstalling numpy-1.22.3:
  Successfully uninstalled numpy-1.22.3
Found existing installation: pandas 1.4.2
Uninstalling pandas-1.4.2:
  Successfully uninstalled pandas-1.4.2
Note: you may need to restart the kernel to use updated packages.


In [10]:
# %pip cache purge

Files removed: 5
Note: you may need to restart the kernel to use updated packages.
